Assistants usage

lasryaric opened the pull request:
Adding assistant usage in the context of Ahuna.
Task [here](https://github.com/orgs/dust-tt/projects/3?pane=issue&itemId=45548914).

Using Redis instead reading directly in the database to achieve O(log(N)) read time performance, vs O(N) using PostgreSQL.
(N being the number of messages or people you are counting over).

See Redis `sorted sets` and the `ZCOUNT` command for more details.

![Screenshot 2023-12-20 at 17 16 02](https://github.com/dust-tt/dust/assets/358965/5a697394-695f-4b4e-a2de-6c5c2dc5ed77)

lasryaric commented on front/lib/api/assistant/conversation.ts:
```diff
@@ -1054,7 +1057,19 @@ export async function* postUserMessage(
+      // Can't use Promise.all() here because populating the data
+      // on the first time would result in a very likely race condition
+      // that would result in counting multiple time each message.
+      // We'll need to add a lock here.
```
Actually this can be solve by checking if the Redis keys exists using `incr` (vs `exists`).
`Incr` is thread safe, so if we get `1`, it means we need to populate, otherwise we can just insert the data. Order (obviously) does not matter.

philipperolet commented on front/lib/api/assistant/agent_usage.ts:
```diff
@@ -0,0 +1,216 @@
+  // One sorted set per agent for counting the number of times the agent has been used.
+  // score is: timestamp of each use of the agent
+  // value: random unique distinct value.
+  const agentUsedCountKey = `agent_usage_count_${agentConfigurationId}`;
```
nit: would call it agent message count, or agent mention count
usage is very close to user, and is ambiguous, so makes the code less readable

philipperolet replied:
I know this is a big renaming since it's in a lot of places :sweat_smile:

philipperolet commented on front/lib/swr.ts:
```diff
@@ -466,6 +467,29 @@ export function useAgentConfigurations({
+  );
+
+  return {
+    usage: data ? data.agentUsage : null,
```
nit: `usage` -> `agentUsage`

spolu replied:
:+1:

philipperolet commented on front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/usage.ts:
```diff
@@ -0,0 +1,76 @@
+      api_error: {
+        type: "app_auth_error",
+        message:
+          "Only the users that are members for the current workspace can access to the workspace's assistants.",
```
```suggestion
          "Only the users that are members for the current workspace can access the workspace's assistants.",
```

philipperolet commented on types/src/front/assistant/agent.ts:
```diff
@@ -132,3 +132,8 @@ export type AgentConfigurationType = {
+
+export type AgentUsageType = {
+  userCount: number;
+  usageCount: number;
```
idem would rename mentionCount or messageCount

philipperolet replied:
especially since the type is also called AgentUsage

philipperolet reviewed (changes requested):
Thanks for taking that up :pray:  The task is more challenging than it seems !
This is a partial review, reviewed everything except `lib/api/assistant/agent_usage.ts` yet because I have a few questions about the approach, see below

 => most of it looks :+1: , left a few coms

philipperolet commented:
About the use of redis & sorted sets to accomplish this, the major upside I see about the method is that it's fast and reduces the load on the DB
The downside is that the related code is more complex and adds side-effect non-functional logic (i.e. post / edit user message have to do something else than posting a message, they have to signal redis). 

Clearly, it's more important to be fast. But wondering if we could have it all? I presumed you already considered other options, curious to explore:
- a way to make the count query fast with double index + ordering?
- Given that we have to signal redis at every post,  options around maintaining a counter directly AgentConfiguration model with timpestamps? 
- async computing of such a counter, to avoid signalling completely (since the count doesn't have to be real-time);

At the time I didn't fully dig into those so I don't know if they're actual serious candidates

So if you're definite that the redis solution is superior, then :ok_hand:  => I'll review as is (and you can explain me when the time comes why the other options were poorer)

lasryaric commented:
Oopsy! Forgot to add proper support for global agent. I will do that later
tonight.

@pr I agree this looks complicated, but I don’t know of any solution for
counting in sub linear time with Postgresql. (And i looked, it’s a known
limitation of postgres).
Given this is in the assistant details, we could just count from the db
when needed, but if we want something fast to be able to display it
everywhere, then I believe this is a good solution.

On Wed, Dec 20, 2023 at 6:16 PM Philippe Rolet ***@***.***>
wrote:

> About the use of redis & sorted sets to accomplish this, the major upside
> I see about the method is that it's fast and reduces the load on the DB
> The downside is that the related code is more complex and adds side-effect
> non-functional logic (i.e. post / edit user message have to do something
> else than posting a message, they have to signal redis).
>
> Clearly, it's more important to be fast. But wondering if we could have it
> all? I presumed you already considered other options, curious to explore:
>
>    - a way to make the count query fast with double index + ordering?
>    - Given that we have to signal redis at every post, options around
>    maintaining a counter directly AgentConfiguration model with timpestamps?
>    - async computing of such a counter, to avoid signalling completely
>    (since the count doesn't have to be real-time);
>
> At the time I didn't fully dig into those so I don't know if they're
> actual serious candidates
>
> So if you're definite that the redis solution is superior, then 👌 => I'll
> review as is (and you can explain me when the time comes why the other
> options were poorer)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dust-tt/dust/pull/2959#issuecomment-1864849458>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACXUNJK7HZXPHG3BCBBPTTYKMMOPAVCNFSM6AAAAABA3SD4RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNRUHA2DSNBVHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

spolu commented on front/lib/api/assistant/agent_usage.ts:
```diff
@@ -0,0 +1,216 @@
+  messageId: ModelId;
+  redis: Awaited<ReturnType<typeof redisClient>> | undefined;
+}) {
+  let needRedisQuit = false;
```
this variable is not really necesssary is it? it's tied to redis being not null ?

spolu replied:
Sorry I now see the logic :+1: LGTM

spolu commented on front/lib/api/assistant/conversation.ts:
```diff
@@ -1054,7 +1055,16 @@ export async function* postUserMessage(
-
+  if (agentMessages.length > 0) {
+    for (const agentMessage of agentMessages) {
+      await signalAgentUsage({
```
Let's not await here as this is not critical path?

spolu commented on front/lib/api/assistant/conversation.ts:
```diff
@@ -1533,6 +1543,20 @@ export async function* editUserMessage(
 
+  if (agentMessages.length > 0) {
+    for (const agentMessage of agentMessages) {
+      await signalAgentUsage({
```
Same here

spolu commented on front/lib/swr.ts:
```diff
@@ -466,6 +467,29 @@ export function useAgentConfigurations({
+  workspaceId: string;
+  agentConfigurationId: string;
+}) {
+  const agentConfigurationsFetcher: Fetcher<GetAgentUsageResponseBody> =
```
Needs renaming

spolu commented on types/src/front/assistant/agent.ts:
```diff
@@ -132,3 +132,8 @@ export type AgentConfigurationType = {
   generation: AgentGenerationConfigurationType | null;
 };
+
+export type AgentUsageType = {
```
Should we add a timeperiod information here even if this is monthly for now always?

spolu reviewed (commented):
Looks great!

spolu commented:
@lasryaric How are global agents not supported? I think they are no?
@philipperolet this isn't that complicated piece of code and we def don't want to rely on SQL for that IMHO. As far as blocking on signaling to redis, I think we can just void instead of await when we do so to have this run completely asyncrhonously :+1:

spolu commented on front/lib/api/assistant/agent_usage.ts:
```diff
@@ -0,0 +1,216 @@
+  userId: string;
+  timestamp: number;
+  messageId: ModelId;
+  redis: Awaited<ReturnType<typeof redisClient>> | undefined;
```
Looks like we always have a redis handy. Maybe make it compulsory for now?

lasryaric commented:
@philipperolet @spolu addressed all your comments and scoped the redis keys to a workspace to isolate the global message and user count to the workspace level.
